nav2_route/src/breadth_first_search.cpp

```diff
@@ -0,0 +1,170 @@
+// Copyright (c) 2023 Joshua Wallace
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "nav2_route/breadth_first_search.hpp"
+
+#include <iostream>
+
+namespace nav2_route
+{
+
+void BreadthFirstSearch::initMotionModel(int x_size, int y_size)
+{
+  max_index_ = x_size * y_size;
+  x_size_ = x_size;
+
+  neighbors_grid_offsets = {-1, +1,
+                            -x_size, +x_size,
+                            -x_size - 1, -x_size + 1,
+                            +x_size - 1, +x_size + 1};
+
+  std::cout << "offset size " << neighbors_grid_offsets.size() << std::endl;
+}
+
+void BreadthFirstSearch::setCostmap(nav2_costmap_2d::Costmap2D *costmap)
+{
+  std::cout << "Set costmap " << std::endl;
+  costmap_ = costmap;
+
+  unsigned int x_size = costmap_->getSizeInCellsX();
+  unsigned int y_size = costmap_->getSizeInCellsY();
+
+  initMotionModel(static_cast<int>(x_size), static_cast<int>(y_size));
+}
+
+void BreadthFirstSearch::setStart(unsigned int mx, unsigned int my) {
+  start_ = getIndex(mx, my, x_size_);
+  states_.insert(std::make_pair(start_, State()));
+}
+
+void BreadthFirstSearch::setGoals(std::vector<unsigned int> mxs, std::vector<unsigned int> mys)
+{
+  goals_.clear();
+  for(unsigned int i = 0; i < (mxs.size() && mys.size()); ++i) {
+    unsigned int index = getIndex(mxs[i], mys[i], x_size_);
+    goals_.push_back(index);
+    states_.insert(std::make_pair(index, State()));
```

Comment:
You end up clearing states first thing in `search`, I think you can remove that here?